jena-text/src/main/java/org/apache/jena/query/text/TextIndexES.java

```diff
@@ -0,0 +1,427 @@
+             * The reason is we want to add a document if it does not exist with the given Subject Id (URI).
+             * But if the document exists with the same Subject Id, we want to do an update to it instead of deleting it and
+             * then creating it with only the latest field values.
+             * This functionality is called Upsert functionality and more can be learned about it here:
+             * https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update.html#upserts
+             */
+
+            //First Search of the field exists or not
+            SearchResponse existsResponse = client.prepareSearch(INDEX_NAME)
+                    .setTypes(docDef.getEntityField())
+                    .setQuery(QueryBuilders.existsQuery(fieldToAdd))
+                    .get();
+            String script;
+            if(existsResponse != null && existsResponse.getHits() != null && existsResponse.getHits().totalHits() > 0) {
+                //This means field already exists and therefore we should append to it
+                script = "ctx._source." + fieldToAdd+".add('"+ fieldValueToAdd + "')";
+            } else {
+                //The field does not exists. so we create one
+                script = "ctx._source." + fieldToAdd+" =['"+ fieldValueToAdd + "']";
+            }
+
+
+
+            UpdateRequest upReq = new UpdateRequest(INDEX_NAME, docDef.getEntityField(), entity.getId())
+                    .script(new Script(script))
+                    .upsert(indexRequest);
+
+            UpdateResponse response = client.update(upReq).get();
+
+            LOGGER.debug("Received the following Update response : " + response + " for the following entity: " + entity);
+
+        } catch(Exception e) {
+            throw new TextIndexException("Unable to Index the Entity in ElasticSearch.", e);
+        }
+
+
+    }
+
+    /**
+     * Delete an entity.
+     * Since we are storing different predicate values within the same indexed document,
+     * deleting the document using entity Id is sufficient to delete all the related contents for a given entity.
+     * @param entity entity to delete
+     */
+    @Override
+    public void deleteEntity(Entity entity) {
+
+        String fieldToRemove = null;
+        String valueToRemove = null;
+        for(String field : docDef.fields()) {
+            if(entity.get(field) != null) {
+                fieldToRemove = field;
+                valueToRemove = (String)entity.get(field);
+                break;
+            }
+        }
+        //First Search of the field exists or not
+        SearchResponse existsResponse = client.prepareSearch(INDEX_NAME)
+                .setTypes(docDef.getEntityField())
+                .setQuery(QueryBuilders.existsQuery(fieldToRemove))
+                .get();
+
+        String script = null;
+        if(existsResponse != null && existsResponse.getHits() != null && existsResponse.getHits().totalHits() > 0) {
```

Comment:
Similar to the update operation, this may cause a race condition if multiple deletes happen simultaneously. Would it be possible to do an atomic delete here, scripted in such a way that there is no error thrown even if the field value to delete doesn't exist in the index?